File: initHMM.R
#' Initializing treeHMM with given parameters
#'
#' @param States A (2 * 1) vector with first element being discrete state value for the cases(or positive) and second element being discrete state value for the controls(or negative) for given treeHMM
#' @param Symbols List containing (M * 1) vectors for discrete values of emissions(where "M" is the possible number of emissions) for each feature variable
#' @param treemat Adjacent Symmetry Matrix that describes the topology of the tree
#' @param startProbs (N * 1) vector containing starting probabilities for the states, where "N" is the possible number of states(Optional). Default is equally probable states
#' @param transProbs (N * N) matrix containing transition probabilities for the states, where "N" is the possible number of states(Optional)
#' @param emissionProbs List of (N * M) matrices containing emission probabilities for the states, for each feature variable(optional). Default is equally probable emissions
#' @return List describing the parameters of treeHMM(pi, alpha, beta)
#' @examples
#' tmat = matrix(c(0,0,1,0,0,0,0,1,0,0,0,0,0,1,1,0,0,0,0,0,0,0,0,0,0),
#'                5,5, byrow= TRUE ) #for "X" (5 nodes) shaped tree
#' states = c("P","N") #"P" represent cases(or positive) and "N" represent controls(or negative)
#' hmmA = initHMM(states,list(c("L","R")), tmat) #one feature with two discrete levels "L" and "R"
#' hmmB = initHMM(states, list(c("X","Y")),tmat, c(0.5,0.5), matrix(c(0.7,0.3,0.3,0.7),2,2))

initHMM = function (States, Symbols, treemat, startProbs = NULL, transProbs = NULL,
                    emissionProbs = NULL)
{
  nStates = length(States)
  nLevel = length(Symbols)
  E=list()
  S = rep((1/nStates),nStates)
  T = 0.5 * diag(nStates) + array(0.5/(nStates), c(nStates,nStates))
  names(S) = States
  dimnames(T) = list(from = States, to = States)
  if (!is.null(startProbs)) {
    S[] = startProbs[]
  }
  if (!is.null(transProbs)) {
    T[, ] = transProbs[, ]
  }
  for(i in 1:nLevel)
  {
    nSymbols=length(Symbols[[i]])
    E[[i]] = array(1/(nSymbols), c(nStates, nSymbols))
    dimnames(E[[i]]) = list(states = States, symbols = Symbols[[i]])
    if (!is.null(emissionProbs[[i]])) {
      E[[i]][, ] = emissionProbs[[i]][, ]
    }
  }
  return(list(States = States, Symbols = Symbols, startProbs = S,
              transProbs = T, emissionProbs = E, adjsym=treemat))
}

File: python/initHMM.py
#!/usr/bin/env python3

# This file initializes the treeHMM with given parameters

# Importing the libraries

import numpy as np
import pandas as pd

# Defining the initHMM function

import pdb
def initHMM(
        states,
        symbols,
        adjacent_symmetry_matrix_values,
        initial_probabilities=None,
        state_transition_probabilities=None,
        emission_probabilities=None):
    """
    Args:
        states: It is a numpy array with first element being discrete state
            value for the cases(or positive) and second element being discrete
            state value for the controls(or negative) for given treeHMM
        symbols: It is a list of numpy array consisting discrete values of
            emissions(where "M" is the possible number of emissions) for each
            feature variable
        adjacent_symmetry_matrix_values: It is the Adjacent Symmetry Matrix that
            describes the topology of the tree
        initial_probabilities: It is a numpy array of shape (N * 1) containing
            initial probabilities for the states, where "N" is the possible
            number of states(Optional). Default is equally probable states
        state_transition_probabilities: It is a numpy array of shape (N * N)
            containing transition probabilities for the states, where "N" is the
            possible number of states(Optional).
        emission_probabilities: It is a list of numpy arrays of shape (N * M)
            containing emission probabilities for the states, for each feature
            variable(optional). Default is equally probable emissions

    Returns:
        hmm: A dictionary describing the parameters of treeHMM
    """

    number_of_states = len(states)
    number_of_levels = len(symbols)
    emission_probabilities_values = list()
    default_initial_probabilities = np.repeat(
        (1 / number_of_states), number_of_states, axis=0)
    default_transition_probabilities = 0.5 * np.identity(number_of_states) + np.ones(shape=(number_of_states, number_of_states)) * (0.5 / number_of_states)
    initial_probabilities_values = dict(
        zip(states, default_initial_probabilities))
    transition_probabilities_values = pd.DataFrame(
        data=default_transition_probabilities, index=states, columns=states)

    if initial_probabilities is not None:
        initial_probabilities_values = dict(zip(states, initial_probabilities))

    if state_transition_probabilities is not None:
        state_transition_probabilities = np.array(
            state_transition_probabilities)
        transition_probabilities_values = pd.DataFrame(
             data=state_transition_probabilities, index=states, columns=states)

    for i in range(number_of_levels):
        number_of_symbols = len(symbols[i])
        emission_probabilities_values.append(np.ones(shape=(number_of_states, number_of_symbols)) * (1 / number_of_symbols))
        emission_probabilities_values[i] = pd.DataFrame(data=emission_probabilities_values[i], index=states, columns=symbols[i])

        if emission_probabilities is not None:
            emission_probabilities_values[i] = pd.DataFrame(data=emission_probabilities, index=states, columns=symbols[i])

    return {
        "states": states,
        "symbols": symbols,
        "initial_probabilities": initial_probabilities_values,
        "state_transition_probabilities": transition_probabilities_values,
        "emission_probabilities": emission_probabilities_values,
        "adjacent_symmetry_matrix": adjacent_symmetry_matrix_values}


def run_an_example():
    """
    sample run for initHMM function
    """
    sample_tree = np.array([0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
                     1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]).reshape(5, 5)
    states = ['P', 'N']
    symbols = [['L', 'R']]
    hmm = initHMM.initHMM(states, symbols, sample_tree)

if __name__ == "__main__":
    run_an_example()
